server/src/main/java/io/crate/metadata/SystemClock.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class SystemClock {
+
+    public static int MICRO_PRECISION = 6; // microseconds
+    public static int MILLI_PRECISION = 3; // milliseconds
+
+    private static final AtomicReference<Clock> CURRENT_CLOCK = new AtomicReference<>(Clock.systemUTC());
+
+    public static void setClock(Clock newClock) {
+        CURRENT_CLOCK.set(newClock);
+    }
+
+    public static void setCurrentMillisFixedUTC(long millis) {
+        setClock(Clock.fixed(Instant.ofEpochMilli(millis), ZoneOffset.UTC));
+    }
+
+    public static void setCurrentMillisSystemUTC() {
+        setClock(Clock.systemUTC());
+    }
+
+    public static Instant currentInstant() {
+        return CURRENT_CLOCK.get().instant();
+    }
+
+    public static long currentTimeMillis() {
+        return toMillis(currentInstant());
+    }
+
+    public static long currentTimeMillis(Instant i) {
+        return toMillis(i);
+    }
+
+    public static long currentTimeMillis(Instant i, int precision) {
+        return withPrecision(toMillis(i), precision, MILLI_PRECISION);
+    }
+
+    public static long microsFromMidnightUTC(Instant i) {
+        return toMicros(i) - (extractDateMillis(i) * 1000L);
+    }
+
+    public static long microsFromMidnightUTC(Instant i, int precision) {
+        return withPrecision(microsFromMidnightUTC(i), precision, MICRO_PRECISION);
+    }
+
+    private static long toMicros(Instant i) {
+        return (i.getEpochSecond() * 1000_000_000L + i.getNano()) / 1000L;
+    }
+
+    private static long toMillis(Instant i) {
+        return (i.getEpochSecond() * 1000_000_000L + i.getNano()) / 1000_000L;
+    }
```

Comment:
Is this broad API surface really necessary? The prior version had a narrower surface, why was the extension necessary?

Reply:
My line of thinking is that all the logic in this class, now, is related to obtaining current time in either micros/millis precision, and then with user defined (additional) precision. For the case of `current_time` the date needs to be removed, leaving the API:

- `public static void setClock(Clock newClock)`: your own clock, maybe you have the need to buy an actual atomic clock and you want to hook it up. 
- `public static void setCurrentMillisFixedUTC(long millis)`: so that you control what is returned, for tests.
- `public static void setCurrentMillisSystemUTC()`: to restore the default system clock.
- `public static Instant currentInstant()`:  This is what the transaction context will end up caching, and is used to produce millis/micros.
- `public static long currentTimeMillis()`:  current millis now, no cached value of instant.
- `public static long currentTimeMillis(Instant i)`: current millis from potentially a cached value of instant.
- `public static long currentTimeMillis(Instant i, int precision)`: as above, with additional user defined precision.
- `public static long microsFromMidnightUTC(Instant i)`: current micros from potentially a cached value of instant. In the code there is no need for `public static long microsFromMidnightUTC()`, so I am not adding surface here.
- `public static long microsFromMidnightUTC(Instant i, int precision)`: as above, but with additional user defined precision.

The rest of the methods are not accessible, they are private.

All clock logic required for the time related functions is put together in this simple class.

Reply:
I think it is pushing too much of what the scalar functions should do into the `SystemClock` class. 
My initial comment suggesting a `Clocks` class was to have a class that is *only* providing the functionality to get a `Clock` instance or an `Instant` instance - and being able to adjust what kind of `Clock` is returned and used. Single responsibility.


The `precision` handling is specific to `current_timestamp` and `current_time` and has no generic use. So I think it doesn't belong into a generic `Clocks` or `SystemClock` class.

Reply:
ok, will amend :)